### PR TITLE
Fix: `eth_call` error handling

### PR DESCRIPTION
**Pull request by canonbrother**
## Summary

- Expect error via `eth_call` with failure request

#### Failure `eth_call` on Defi
```sh
curl http://localhost:19551/ \
  -H 'content-type:application/json' \
  --data-binary \
  '{
    "jsonrpc":"2.0",
    "id":"2",
    "method":"eth_call",
    "params":[{"from":"0x9b8a4af42140d8a4c153a822f02571a1dd037e89", "to":"0x002C2DE24B9120c1FED34DB628091653e9E026B5", "data":"0xdc8886230000000000000000000000000000000000000000000000000000000000000000", "value":"0x0", "gas":"0x7530", "gasPrice": "0x37e11d600"}]
  }'
{"jsonrpc":"2.0","result":"0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001c56616c7565206d7573742062652067726561746572207468616e203000000000","id":"2"}%
```

btw we can still decode the error message from bytes by following
```js
const res = await wallet.call(tx)
console.log('res: ', res)
const msg = ethers.utils.toUtf8String('0x' + res.substring(138))
console.log('msg: ', msg)
```    

#### Failure `eth_call` on Ethereum

```sh
curl https://goerli.drpc.org/ \
  -H 'content-type:application/json' \
  --data-binary \
  '{
    "jsonrpc":"2.0",
    "id":"2",
    "method":"eth_call",
    "params":[{"from":"0x148772F29058DcC772613260b078dCa8C14afF6c", "to":"0x7c2e5e5c473c94b060e2f37e6b9787681d52b4c0", "data":"0xdc8886230000000000000000000000000000000000000000000000000000000000000000", "value":"0x0", "gas":"0x7530", "gasPrice": "0x37e11d600"}]
  }'
{"id":"2","jsonrpc":"2.0","error":{"message":"execution reverted: Value must be greater than 0","code":3,"data":"0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001c56616c7565206d7573742062652067726561746572207468616e203000000000"}}%
```

## RPCs

- eth_call

### CODE REVIEW

**Comment by canonbrother** [comment]
```
{"jsonrpc":"2.0","error":{"code":-32001,"message":"Custom error: VM Exception while processing transaction: revert Value must be greater than 0"},"id":"2"}
```

**Comment by canonbrother** [comment]
{"jsonrpc":"2.0","error":{"code":-32603,"message":"execution reverted: Value must be greater than 0","data":"0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001c56616c7565206d7573742062652067726561746572207468616e203000000000"},"id":"2"}

**Comment by canonbrother** [comment]
tested on geth
```
> eth.call({from:"0x148772f29058dcc772613260b078dca8c14aff6c",to:"0x00212c06790b4373b47b3e697a05ab151550612d",data:"0xdc8886230000000000000000000000000000000000000000000000000000000000000000",value:"0x0",gas:"0x5ad20",gasPrice:"0x2540be400"})
WARN [10-27|12:20:12.963] Served eth_call                          reqid=24 duration="847.417µs" err="execution reverted: Value must be greater than 0" errdata=0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001c56616c7565206d7573742062652067726561746572207468616e203000000000
Error: execution reverted: Value must be greater than 0
	at web3.js:6365:9(39)
	at send (web3.js:5099:62(29))
	at <eval>:1:9(15)
```

File: lib/ain-grpc/src/rpc/eth.rs [hunk 1]
```diff
@@ -357,7 +358,11 @@ impl MetachainRPCServer for MetachainRPCModule {
                 block_number,
             })
             .map_err(RPCError::EvmError)?;
-        Ok(Bytes(data))
+
+        match error_on_execution_failure(&exit_reason, &data) {
+            Ok(_) => Ok(Bytes(data)),
+            Err(e) => Err(e),
+        }
```
Comment (sieniven) [inline lib/ain-grpc/src/rpc/eth.rs:365]: Should consider passing the entire TxResponse instance into the error_on_execution_failure method, which returns the RpcResult<Bytes> itself to skip the entire match code block

File: lib/ain-grpc/src/rpc/eth.rs [hunk 2]
```diff
@@ -1229,3 +1234,55 @@ fn sign(
...
+fn error_on_execution_failure(reason: &ExitReason, data: &[u8]) -> RpcResult<()> {
+    match reason {
+        ExitReason::Succeed(_) => Ok(()),
+        ExitReason::Error(err) => {
+            if *err == ExitError::OutOfGas {
+                return Err(internal_err("out of gas"));
+            }
+            Err(internal_err_with_data(format!("evm error: {err:?}"), &[]))
+        }
+        ExitReason::Revert(_) => {
```
Comment (sieniven) [inline lib/ain-grpc/src/rpc/eth.rs:1247]: Revert arm does not contain a field, does not need a placeholder

File: lib/ain-grpc/src/rpc/eth.rs [hunk 3]
```diff
@@ -357,7 +358,11 @@ impl MetachainRPCServer for MetachainRPCModule {
                 block_number,
             })
             .map_err(RPCError::EvmError)?;
-        Ok(Bytes(data))
+
+        match error_on_execution_failure(&exit_reason, &data) {
```
[the remark this answers is not in the document]
Comment (canonbrother) [reply lib/ain-grpc/src/rpc/eth.rs:362]: was merge conflicts issue.. missing destruct from exit_reason.. fixed
https://github.com/DeFiCh/ain/pull/2620/commits/2d48575f80f0f4e4c8f36ce043b2fdf1d5baf333

File: test/functional/feature_evm_contracts.py [hunk 4]
```diff
@@ -440,6 +441,46 @@ def non_payable_proxied_contract(self):
...
+
+        # should be no error
+        contract.functions.gt0(1).call()
+
+        # should throw error from `call`
+        try:
+            contract.functions.gt0(0).call()
+            raise AssertionError("should not reach here")
+        except web3.exceptions.ContractLogicError as e:
+            assert_equal(e.message, "execution reverted")
```
[the remark this answers is not in the document]
Comment (canonbrother) [reply test/functional/feature_evm_contracts.py]: fixed
https://github.com/DeFiCh/ain/pull/2620/commits/c222b82fffb57f52fc3cac07386b261aabb1e90d

File: lib/ain-grpc/src/rpc/eth.rs (hunk 2 above)
Comment (canonbrother) [reply lib/ain-grpc/src/rpc/eth.rs:1247]: that got something `Revert(ExitRevert)` but here doesn't need it
https://docs.rs/evm-core/latest/src/evm_core/error.rs.html#30

File: lib/ain-grpc/src/rpc/eth.rs (hunk 1 above)
Comment (canonbrother) [reply lib/ain-grpc/src/rpc/eth.rs:366]: hmm.. not really sure what you mean.. but ExitReason match is still required

**Pull request merged**